Templating important classes

benjaminhuth opened the pull request:
The discussed changes from issue #102.

dguest commented:
Hi @benjaminhuth, what's left to do here? Overall this looks very nice, I'm trying to decide when we should merge it.

benjaminhuth commented:
Hi, 
from my perspective nothing. But I thought there would be for sure some things to change in such a large update, so made a WIP PR. I think I misunderstood the concept of a WIP PR a bit...

dguest commented:
Hmm, github changed their interface and I didn't notice that the CI pipelines are all broken now... I should fix these before we make such a big change.

dguest commented:
I'm trying to figure out how to run the Travis pipelines on this merge request. It seems like something broke when microsoft started pushing actions and apps.

dguest commented:
Hi @benjaminhuth, can you try to rebase this branch to the current master? I just managed to get the CI pipeline running again, once you rebase it should work for you as well.

dguest commented on include/lwtnn/LightweightGraph.hh:
```diff
@@ -42,19 +42,21 @@
 
-#include "lightweight_network_config.hh"
+#include <memory>
+#include <lwtnn/lightweight_network_config.hh>
```
This might be a bit of a nitpick, but it seems like changing these to `<...>` from `"..."` is adding commit noise. Is there a good reason to make this change?

dguest commented on include/lwtnn/Stack.hh:
```diff
@@ -1,360 +1,52 @@
+#ifndef LWTNN_STACK_HH
+#define LWTNN_STACK_HH
+
+#include <lwtnn/generic/Stack.hh>
```
Same comment here with the angle brackets include. I'd rather keep it consistently using quotes for anything in this library, angle brackets for system libraries.

dguest commented on include/lwtnn/generic/Graph.hh:
```diff
@@ -0,0 +1,173 @@
+#include <unordered_map>
+#include <set>
+
+#include <lwtnn/generic/eigen_typedefs.hh>
+#include <lwtnn/generic/Source.hh>
+#include <lwtnn/NNLayerConfig.hh>
```
Also here with the angle bracket includes

dguest commented on include/lwtnn/generic/Graph.txx:
```diff
@@ -1,60 +1,71 @@
-#include "lwtnn/Graph.hh"
-#include "lwtnn/Exceptions.hh"
-#include "lwtnn/Stack.hh"
+#include <lwtnn/generic/Stack.hh>
+#include <lwtnn/generic/Graph.hh>
+#include <lwtnn/Exceptions.hh>
 
```
This might be the worst offender.

dguest commented on include/lwtnn/generic/InputPreprocessor.hh:
```diff
@@ -1,50 +1,59 @@
-
-  using Eigen::VectorXd;
-  using Eigen::MatrixXd;
+#include <lwtnn/lightweight_network_config.hh>
+#include <lwtnn/Exceptions.hh>
+#include <lwtnn/generic/eigen_typedefs.hh>
```
again with the angle brackets

dguest commented on include/lwtnn/generic/InputPreprocessor.txx:
```diff
@@ -1,14 +1,21 @@
-#include "lwtnn/InputPreprocessor.hh"
+#include <lwtnn/generic/InputPreprocessor.hh>
```
angle brackets

dguest commented on include/lwtnn/generic/LightweightNeuralNetwork.txx:
```diff
@@ -0,0 +1,114 @@
+#include <lwtnn/generic/LightweightNeuralNetwork.hh>
+#include <lwtnn/generic/InputPreprocessor.hh>
+#include <lwtnn/generic/Stack.hh>
+#include <lwtnn/generic/eigen_typedefs.hh>
```
angle brackets include

dguest commented on include/lwtnn/generic/Source.hh:
```diff
@@ -0,0 +1,54 @@
+#ifndef LWTNN_GENERIC_SOURCE_HH
+#define LWTNN_GENERIC_SOURCE_HH
+
+#include <lwtnn/generic/eigen_typedefs.hh>
```
another one

dguest commented on include/lwtnn/generic/Stack.txx:
```diff
@@ -0,0 +1,795 @@
+#include <lwtnn/generic/Stack.hh>
```
use quotes here

dguest commented on src/LightweightGraph.cxx:
```diff
@@ -1,157 +1,43 @@
-#include "lwtnn/InputPreprocessor.hh"
-#include "lwtnn/Graph.hh"
-#include <Eigen/Dense>
+#include <lwtnn/LightweightGraph.hh>
+#include <lwtnn/generic/LightweightGraph.hh>
+#include <lwtnn/generic/InputPreprocessor.hh>
+#include <lwtnn/generic/Graph.hh>
+#include <lwtnn/generic/eigen_typedefs.hh>
```
another example of changing the quote style. `<Eigen/Dense>` should remain as it is because it's an external library (although here I guess it's just removed).

dguest commented on src/LightweightNeuralNetwork.cxx:
```diff
@@ -1,6 +1,7 @@
-#include "lwtnn/LightweightNeuralNetwork.hh"
-#include "lwtnn/InputPreprocessor.hh"
-#include "lwtnn/Stack.hh"
+#include <lwtnn/LightweightNeuralNetwork.hh>
+#include <lwtnn/generic/LightweightNeuralNetwork.hh>
+#include <lwtnn/generic/InputPreprocessor.hh>
+#include <lwtnn/generic/Stack.hh>
```
here too

dguest reviewed (commented):
Just to add another general nitpick: it would be nice if you used a template extension that github recognized as C++. I believe `*.tcc` should work.

Otherwise my main request is that you try to keep the use of `#include` consistent: anything referencing this library should use quotes, while external libraries should use angle brackets.

benjaminhuth commented:
Sorry for the delay, I was on vacation last week.

> Just to add another general nitpick: it would be nice if you used a template extension that github recognized as C++. I believe `*.tcc` should work.
> 
> Otherwise my main request is that you try to keep the use of `#include` consistent: anything referencing this library should use quotes, while external libraries should use angle brackets.

Makes sense to be consistent with these things. I changed the include notion and the file extension to `*.tcc`. Additionally to that, I added include guards to the template implementation files.

The rebase is also done, so the PR should be ready for merge, right?

dguest commented:
Hi @benjaminhuth, do you have any objection to squashing this MR? It seems to do exactly one thing (even if it touches a lot of code), but if you wanted to keep it as multiple commits for some reason we should discuss how that would be done.

benjaminhuth commented:
I didn't know yet that it is possible to squash the commits somehow... But for me its is perfectly ok to do it as one single commit!

dguest commented:
Alright, merged! Thanks for all the work, let's see if anything breaks!

dguest commented on include/lwtnn/generic/Stack.tcc:
```diff
@@ -0,0 +1,800 @@
+
+  template<typename T>
+  T nn_relu( T x) {
+    if (std::isnan(static_cast<double>(x))) return x;
```
Hi @benjaminhuth!

I was about to push a new release into our reconstruction code, and I got curious about this line. It _probably_ won't be a problem, but why is it needed? I assume this was put in place to work with autodiff? 

I thought the more generic thing might be to use the same trick that we use in the above functions, e.g.

```
using std::isnan;
if (isnan(x)) return x;
...
```
which should let anyone overload `isnan` however they want. Do you think that would work?

(It turns out some people are interested in using this version of lwtnn with `T = float` to try to reduce memory consumption. We also just started running a lot more NNs in our reconstruction code, so I'm trying to speed things up a bit as well.)

benjaminhuth replied:
Hi @dguest,
Yeah that's a leftover of my autodiff-experiments. I must have missed that during my cleanup... The problem was that autodiff had no support for `std::isnan`, and this was just a quick workaround. The better way of doing this is of course the 
```c++
using std::foo;
foo(...);
```
I also used this for `std::sqrt`, etc... However, I think it would be better to do this `using` consistently for all these functions, and if I will use autodiff again someday with LWTNN, I'll just implement the correct overloads...

dguest replied:
Hi @benjaminhuth, thanks for the quick reply! I'm about to tag a new release anyway (there were a few other things added recently that we need in ATLAS reconstruction) so I can make this change myself if you think `using std::foo; foo(...);` will accommodate autodiff.
